Userland/Libraries/LibAudio/PlaybackStreamAudioUnit.cpp

```diff
@@ -0,0 +1,396 @@
+    }
+
+private:
+    AudioState(AudioStreamBasicDescription description, AudioTaskQueue task_queue, PlaybackStream::AudioDataRequestCallback data_request_callback, OutputState initial_output_state)
+        : m_description(description)
+        , m_task_queue(move(task_queue))
+        , m_paused(initial_output_state == OutputState::Playing ? Paused::No : Paused::Yes)
+        , m_data_request_callback(move(data_request_callback))
+    {
+    }
+
+    static OSStatus on_audio_unit_buffer_request(void* user_data, AudioUnitRenderActionFlags*, AudioTimeStamp const* time_stamp, UInt32 element, UInt32 frames_to_render, AudioBufferList* output_buffer_list)
+    {
+        VERIFY(element == AUDIO_UNIT_OUTPUT_BUS);
+        VERIFY(output_buffer_list->mNumberBuffers == 1);
+
+        auto& state = *static_cast<AudioState*>(user_data);
+
+        VERIFY(time_stamp->mFlags & kAudioTimeStampSampleTimeValid);
+        auto sample_time_seconds = time_stamp->mSampleTime / state.m_description.mSampleRate;
+
+        auto last_sample_time = static_cast<i64>(sample_time_seconds * 1000.0);
+        state.m_last_sample_time.store(last_sample_time);
+
+        if (auto result = state.m_task_queue.dequeue(); result.is_error()) {
+            VERIFY(result.error() == AudioTaskQueue::QueueStatus::Empty);
+        } else {
+            auto task = result.release_value();
+            OSStatus error = noErr;
+
+            switch (task.type) {
+            case AudioTask::Type::Play:
+                state.m_paused = Paused::No;
+                break;
+
+            case AudioTask::Type::Pause:
+                state.m_paused = Paused::Yes;
+                break;
+
+            case AudioTask::Type::PauseAndDiscard:
+                error = AudioUnitReset(state.m_audio_unit, kAudioUnitScope_Global, AUDIO_UNIT_OUTPUT_BUS);
+                state.m_paused = Paused::Yes;
+                break;
+
+            case AudioTask::Type::Volume:
+                VERIFY(task.data.has_value());
+                error = AudioUnitSetParameter(state.m_audio_unit, kHALOutputParam_Volume, kAudioUnitScope_Global, 0, static_cast<float>(*task.data), 0);
+                break;
+            }
+
+            if (error == noErr)
+                task.resolve(Duration::from_milliseconds(last_sample_time));
+            else
+                task.reject(error);
+        }
+
+        Bytes output_buffer {
+            reinterpret_cast<u8*>(output_buffer_list->mBuffers[0].mData),
+            output_buffer_list->mBuffers[0].mDataByteSize
+        };
+
+        if (state.m_paused == Paused::No) {
+            auto written_bytes = state.m_data_request_callback(output_buffer, PcmSampleFormat::Float32, frames_to_render);
+
+            if (written_bytes.is_empty())
+                state.m_paused = Paused::Yes;
+        }
+
+        if (state.m_paused == Paused::Yes)
+            output_buffer.fill(0);
+
+        return noErr;
+    }
+
+    AudioComponentInstance m_audio_unit { nil };
```

Comment:
I see you moved this class from a `.mm` file to a `.cpp` file when swapping to AudioUnit from AudioToolbox. Is it still correct to use `nil` here, or should it be `nullptr` ?

Reply:
Ah yeah. There's no objc-ness in this file anymore so those can change to be more C++-y (`nil` is just a `#define` to `nullptr` since C++11)